In [1]:
import pandas as pd
import gzip
import os
import json
pd.set_option('display.max_columns', None)

/Users/filippo/.pyenv/versions/3.8.7/lib/python3.8/site-packages/pandas/compat/__init__.py:97: UserWarning: Could not import the lzma module. Your installed Python is incomplete. Attempting to use lzma compression will result in a RuntimeError.
  warnings.warn(msg)


In [3]:
def plain_dict(d):
    tmp_d = {}
    for key in d.keys():
        
        if type(d[key]) is dict:
            tmp_sub_dict = plain_dict(d[key])
            for key_tmp in tmp_sub_dict.keys():
                tmp_d[key + "__" + key_tmp] = tmp_sub_dict[key_tmp]
        
        elif type(d[key]) is list or type(d[key]) is set or type(d[key]) is frozenset:
            tmp_d[key] = len(d[key])
        else:
            tmp_d[key] = d[key]
    return tmp_d

In [31]:
plain_dict({"s": "s", "a" : {"s": "aa", "k": {"kk": "llll"}}})

{'s': 's', 'a': {'s': 'aa', 'k': {'kk': 'llll'}}}
else: s
a
{'s': 'aa', 'k': {'kk': 'llll'}}
else: s
k
{'kk': 'llll'}
else: kk


{'s': 's', 'a__s': 'aa', 'a__k__kk': 'llll'}

In [5]:
FOLDER = './ni'
resources = []
for file in os.listdir(FOLDER):
        
        if not os.path.isfile(os.path.join(FOLDER, file)) or not file.endswith('.gz'):
            continue
        
        with gzip.open(os.path.join(FOLDER, file), 'rb') as f:
            # each file line contain a resource
            # variables ending with _n refer with element obtained by Neo4j insert
            resources.extend([plain_dict(json.loads(line)) for line in f.readlines()])

In [48]:
print(resources[0])

{'author': 3, 'bestaccessright__code': 'c_abf2', 'bestaccessright__label': 'OPEN', 'bestaccessright__scheme': 'http://vocabularies.coar-repositories.org/documentation/access_rights/', 'collectedfrom': 1, 'context': 1, 'contributor': 0, 'country': 0, 'coverage': 0, 'dateofcollection': '2020-11-18T17:06:01.948Z', 'description': 0, 'format': 0, 'geolocation': 0, 'id': '50|NeuroVault__::1b5230e3ea6d7123847ad55a1e06fffd', 'instance': 1, 'language__code': 'UNKNOWN', 'language__label': 'UNKNOWN', 'lastupdatetimestamp': 1608740451078, 'maintitle': 'Peri-hand space representation in the absence of a hand – Evidence from congenital one-handers', 'originalId': 1, 'pid': 0, 'publicationdate': '2017-09-04', 'subjects': 1, 'type': 'dataset'}


In [49]:
df = pd.DataFrame(resources)

In [59]:
df

author bestaccessright__code bestaccessright__label  \
0          3.0                c_abf2                   OPEN   
1          4.0                c_abf2                   OPEN   
2          5.0                   NaN                    NaN   
3          1.0                   NaN                    NaN   
4          5.0                   NaN                    NaN   
...        ...                   ...                    ...   
209264     2.0                c_abf2                   OPEN   
209265    22.0                c_abf2                   OPEN   
209266     1.0                c_abf2                   OPEN   
209267     3.0                c_abf2                   OPEN   
209268     0.0                c_abf2                   OPEN   

                                  bestaccessright__scheme  collectedfrom  \
0       http://vocabularies.coar-repositories.org/docu...              1   
1       http://vocabularies.coar-repositories.org/docu...              1   
2                                                     NaN              1   
3                                                     NaN              1   
4                                                     NaN              1   
...                                                   ...            ...   
209264  http://vocabularies.coar-repositories.org/docu...              2   
209265  http://vocabularies.coar-repositories.org/docu...              3   
209266  http://vocabularies.coar-repositories.org/docu...              2   
209267  http://vocabularies.coar-repositories.org/docu...              2   
209268  http://vocabularies.coar-repositories.org/docu...              1   

        context  contributor  country  coverage          dateofcollection  \
0             1            0      0.0         0  2020-11-18T17:06:01.948Z   
1             1            0      0.0         0  2020-11-18T17:07:17.748Z   
2             1            0      0.0         0  2018-10-28T00:39:04.337Z   
3             1            0      0.0         0  2018-10-28T00:39:04.337Z   
4             1            0      0.0         0  2018-10-28T00:39:04.337Z   
...         ...          ...      ...       ...                       ...   
209264        1           16      0.0         0  2018-10-28T00:39:04.337Z   
209265        2            0      0.0         0  2020-11-08T23:27:58.067Z   
209266        1            0      0.0         0  2020-01-31T23:06:18.565Z   
209267        1            0      0.0         0   2020-01-31T23:05:47.93Z   
209268        1            0      0.0         0                             

        description  format  geolocation  \
0                 0       0          0.0   
1                 1       0          0.0   
2                 0       0          0.0   
3                 1       0          0.0   
4                 0       0          0.0   
...             ...     ...          ...   
209264            1       0          NaN   
209265            1       0          NaN   
209266            1       0          NaN   
209267            1       0          NaN   
209268            1       0          NaN   

                                                       id  instance  \
0       50|NeuroVault__::1b5230e3ea6d7123847ad55a1e06fffd       1.0   
1       50|NeuroVault__::aab3238922bcc25a6f606eb525ffdc56       1.0   
2       50|datacite____::3a3d161270ebecb0e88286a79b2604fa       1.0   
3       50|datacite____::a830848fc3225c99d594becaa5dcbeb5       1.0   
4       50|datacite____::f62842400cd4a6d409f5f76318d3fa7c       1.0   
...                                                   ...       ...   
209264  50|dedup_wf_001::1144aebf1a47a0975b035a2c1b90c15e       2.0   
209265  50|dedup_wf_001::53027b2ed4dbe76f4220035b15cf3265       4.0   
209266  50|dedup_wf_001::855a903073747e34921ed0c1d4ddf50d       2.0   
209267  50|dedup_wf_001::76949a708a4f8c859edff0d44c4cc93b       3.0   
209268  50|r37980778c78::877210419dd492b19ab269769aa47778       2.0   

       language__code language_

In [52]:
print(len(df))

209269


In [56]:
df.describe()

author  collectedfrom        context    contributor  \
count  209223.000000  209269.000000  209269.000000  209269.000000   
mean        5.318263       3.947360       1.011899       0.316463   
std         8.061297       2.334681       0.115556       2.216044   
min         0.000000       1.000000       1.000000       0.000000   
25%         2.000000       2.000000       1.000000       0.000000   
50%         4.000000       3.000000       1.000000       0.000000   
75%         7.000000       5.000000       1.000000       0.000000   
max       967.000000      65.000000       6.000000     343.000000   

             country  coverage    description         format  geolocation  \
count  209179.000000  209269.0  209269.000000  209269.000000  7255.000000   
mean        0.184933       0.0       0.956018       0.143844     0.003032   
std         0.451196       0.0       0.531997       0.421596     0.072314   
min         0.000000       0.0       0.000000       0.000000     0.000000   
25%         0.000000       0.0       1.000000       0.000000     0.000000   
50%         0.000000       0.0       1.000000       0.000000     0.000000   
75%         0.000000       0.0       1.000000       0.000000     0.000000   
max        15.000000       0.0     109.000000      19.000000     3.000000   

            instance  lastupdatetimestamp    originalId            pid  \
count  209236.000000         2.091100e+05  209236.00000  209269.000000   
mean        3.771665         1.591775e+12       3.44261       1.758469   
std         4.011212         2.163189e+10       3.90801       2.465263   
min         1.000000         1.502380e+12       1.00000       0.000000   
25%         2.000000         1.574363e+12       2.00000       1.000000   
50%         3.000000         1.608740e+12       3.00000       1.000000   
75%         5.000000         1.608747e+12       4.00000       3.000000   
max       438.000000         1.608750e+12     589.00000     199.000000   

            subjects      projects  contactgroup  contactperson    tool  \
count  209269.000000  18121.000000        1610.0    1610.000000  1610.0   
mean       18.767056      2.136747           0.0       0.000621     0.0   
std        12.649303      2.929612           0.0       0.024922     0.0   
min         0.000000      1.000000           0.0       0.000000     0.0   
25%        14.000000      1.000000           0.0       0.000000     0.0   
50%        17.000000      1.000000           0.0       0.000000     0.0   
75%        21.000000      2.000000           0.0       0.000000     0.0   
max      1741.000000    182.000000           0.0       1.000000     0.0   

       documentationUrl  
count             423.0  
mean                0.0  
std                 0.0  
min                 0.0  
25%                 0.0  
50%                 0.0  
75%                 0.0  
max                 0.0

In [57]:
df.astype(object).describe()

author bestaccessright__code bestaccessright__label  \
count   209223.0                206835                 206835   
unique     189.0                     4                      4   
top          3.0                c_abf2                   OPEN   
freq     30028.0                107699                 107699   

                                  bestaccessright__scheme  collectedfrom  \
count                                              206835         209269   
unique                                                  1             32   
top     http://vocabularies.coar-repositories.org/docu...              2   
freq                                               206835          60161   

        context  contributor   country  coverage dateofcollection  \
count    209269       209269  209179.0    209269           207208   
unique        5           73      11.0         1           106828   
top           1            0       0.0         0                    
freq     206934       189833  174773.0    209269            35100   

        description  format  geolocation  \
count        209269  209269       7255.0   
unique           23       9          4.0   
top               1       0          0.0   
freq         182365  183817       7240.0   

                                                       id  instance  \
count                                              209269  209236.0   
unique                                             209269     127.0   
top     50|doiboost____::bd99ba4f1ee46044933afa57bccdc1cc       2.0   
freq                                                    1   83982.0   

       language__code language__label  lastupdatetimestamp        maintitle  \
count          209269          209269         2.091100e+05           206308   
unique             41              41         1.045530e+05           204084   
top               und    Undetermined         1.608747e+12  The EEG society   
freq           136746          136746         1.007110e+05               27   

        originalId     pid publicationdate  subjects         type  \
count     209236.0  209269          209189    209269       209269   
unique       123.0      97           51715       198            4   
top            2.0       1      2015-01-01        16  publication   
freq       64679.0  134315            2722     19208       199789   

          publisher version     size  projects embargoenddate  contactgroup  \
count        186156    1795      769   18121.0           3511        1610.0   
unique         4240      16      653      42.0            710           1.0   
top     Elsevier BV    None  0 Bytes       1.0                          0.0   
freq          40990    1699       80    9991.0           2394        1610.0   

        contactperson    tool  documentationUrl programmingLanguage  \
count          1610.0  1610.0             423.0                 423   
unique            2.0     1.0               1.0                   1   
top               0.0     0.0               0.0                       
freq           1609.0  1610.0             423.0                 423   

       container__edition container__ep container__iss container__issnLinking  \
count               26304        108960          26304                  26304   
unique                  1          7337            232                      1   
top                                                                             
freq                26304         19128          17416                  26304   

       container__issnOnline container__issnPrinted container__name  \
count                  77807                 119881          143214   
unique                  5374                   7337           13513   
top                                                        PLoS ONE   
freq                   15559                   8117            4669   

       container__sp container__vol container__conferencedate  \
count         122426         127734                      23

## Riflessioni
967 autori in una risorsa mi sembrano un po' eccessivi, è possibile? Si è un caso normale tra i fisici
I contributor, country, coverage, description, format, geolocation,  sono praticamente sempre 0   
Molti valori compaiono davvero poche 


## Author with orc id

In [6]:
FOLDER = './ni'
authors = []
for file in os.listdir(FOLDER):
        
        if not os.path.isfile(os.path.join(FOLDER, file)) or not file.endswith('.gz'):
            continue
        
        with gzip.open(os.path.join(FOLDER, file), 'rb') as f:
            # each file line contain a resource
            # variables ending with _n refer with element obtained by Neo4j insert
            for line in f.readlines():
                resource = json.loads(line)
                authors_elem = resource.get('author')
                if authors_elem:
                    authors.extend([plain_dict(a) for a in authors_elem])

In [7]:
authors[0]

{'fullname': 'Maimon-Mor, Roni O.',
 'name': 'Roni O.',
 'rank': 1,
 'surname': 'Maimon-Mor'}

In [8]:
authors_df = pd.DataFrame(authors)

In [10]:
authors_df.astype(object).describe() # considerando 48K orchid su forse 426K di authori unici (nomi diversi)

fullname    name     rank surname pid__id__scheme  \
count            1112703  117662  1112703  117655           48329   
unique            426568   25766      967   36978               1   
top     Vince D. Calhoun                1                   orcid   
freq                 596    1767   207537    1761           48329   

             pid__id__value pid__provenance__provenance pid__provenance__trust  
count                 48329                       48329                  48329  
unique                13035                           1                      1  
top     0000-0002-4720-8867                   Harvested                   0.91  
freq                    275                       48329                  48329